lib/charms/microcluster_token_distributor/v0/token_distributor.py

```diff
@@ -115,6 +174,108 @@ def _call_cluster_command(self, *args):
+        """Add data into the units databag in a mirrorable form."""
+        for k, v in data.items():
+            relation.data[self.charm.unit][self._to_mirror_key(k)] = v
+
+    def find_value(self, relation, key, keep_empty=True):
+        """Find corresponding value in mirror.
+
+        Return false if value not found, if keep_empty is enabled 'empty' an
+        empty value will be accepted.
+        """
+        values = [
+            value
+            for mirror in self.find_mirrors(relation)
+            if (value := relation.data[mirror].get(self._to_mirror_key(key)))
+            and (keep_empty or value != "empty")
+        ]
+
+        if len(values) == 0:
+            return False
+
+        return corroborate(values, default="empty")
+
+    def any_data_exists(self, relation):
+        """Check if there is any non empty value on the remote side of the mirror."""
+        for unit in relation.units:
+            for k in relation.data[unit].keys():
+                if k.startswith(MIRROR_PREFIX) and relation.data[unit][k] != "empty":
+                    return True
+        return False
+
+    def find_mirrors(self, relation) -> list[ops.Unit]:
+        """Find all remote mirror units."""
+        relation_data = relation.data
+        distributor_mirrors = [
+            unit
+            for unit in relation.units
+            if (mirror := relation_data[unit].get("mirror")) and mirror == "up"
+        ]
+        return distributor_mirrors
+
+    def get_relevant_mirror_data(self, relation, keep_empty=True) -> dict[str, str]:
+        """Return data in the mirror, where the key is of the form mirror-key.
+
+        Return a dictionary of the relevant data in the mirror, with the mirror
+        prefix stripped from the key.
+        If keep_empty is true treat a value of empty as a valid value.
+        """
+        relation_data = relation.data
+        data = {}
+        for mirror in self.find_mirrors(relation):
+            mirror_data = relation_data[mirror]
+            for mirror_key in mirror_data.keys():
+                logger.info(mirror_key)
+                if not mirror_key.startswith(MIRROR_PREFIX):
+                    continue
+                if (not keep_empty) and mirror_data[mirror_key] == "empty":
+                    continue
+
+                key = mirror_key[len(MIRROR_PREFIX) :]
+                logger.info(key)
+                # create list in case it exists multiple times
+                if key in data:
+                    data[key].append(mirror_data[mirror_key])
+                else:
+                    data[key] = [mirror_data[mirror_key]]
+
+        for key in data.keys():
+            data[key] = corroborate(data[key], default="empty")
+
+        return data
+
+    # this will generate tokens for any relevant data with the value empty
+    # however discerning a hostname from a key without just checking if its of
+    # the standard machine charm unit hostname format is awkward
```

Comment:
nit: Would it make sense to move the above explanation to the function's docstring?